test-integ/topoutil/asserter.go

```diff
@@ -0,0 +1,312 @@
+	service *topology.Service,
+	port int,
+	path string,
+	expectedResHeader map[string]string,
+) {
+	t.Helper()
+	require.True(t, port > 0)
+
+	node := service.Node
+	ip := node.LocalAddress()
+	addr := fmt.Sprintf("%s:%d", ip, port)
+
+	client := a.mustGetHTTPClient(t, node.Cluster)
+	libassert.HTTPServiceEchoesResHeaderWithClient(t, client, addr, path, expectedResHeader)
+}
+
+func (a *Asserter) HTTPStatus(
+	t *testing.T,
+	service *topology.Service,
+	port int,
+	status int,
+) {
+	t.Helper()
+	require.True(t, port > 0)
+
+	node := service.Node
+	ip := node.LocalAddress()
+	addr := fmt.Sprintf("%s:%d", ip, port)
+
+	client := a.mustGetHTTPClient(t, node.Cluster)
+
+	url := "http://" + addr
+
+	retry.RunWith(&retry.Timer{Timeout: 30 * time.Second, Wait: 500 * time.Millisecond}, t, func(r *retry.R) {
+		resp, err := client.Get(url)
+		if err != nil {
+			r.Fatalf("could not make request to %q: %v", url, err)
+		}
+		defer resp.Body.Close()
+		if resp.StatusCode != status {
+			r.Fatalf("expected status %d, got %d", status, resp.StatusCode)
+		}
+	})
+}
+
+// asserts that the service sid in cluster and exported by peer localPeerName is passing health checks,
+func (a *Asserter) HealthyWithPeer(t *testing.T, cluster string, sid topology.ServiceID, peerName string) {
+	t.Helper()
+	cl := a.mustGetAPIClient(t, cluster)
+	retry.RunWith(&retry.Timer{Timeout: time.Minute * 1, Wait: time.Millisecond * 500}, t, func(r *retry.R) {
+		svcs, _, err := cl.Health().Service(
+			sid.Name,
+			"",
+			true,
+			utils.CompatQueryOpts(&api.QueryOptions{
+				Partition: sid.Partition,
+				Namespace: sid.Namespace,
+				Peer:      peerName,
+			}),
+		)
+		require.NoError(r, err)
+		assert.GreaterOrEqual(r, len(svcs), 1)
+	})
+}
+
+func (a *Asserter) UpstreamEndpointHealthy(t *testing.T, svc *topology.Service, upstream *topology.Upstream) {
+	t.Helper()
+	node := svc.Node
+	ip := node.LocalAddress()
+	port := svc.EnvoyAdminPort
+	addr := fmt.Sprintf("%s:%d", ip, port)
+
+	client := a.mustGetHTTPClient(t, node.Cluster)
+	libassert.AssertUpstreamEndpointStatusWithClient(t,
+		client,
+		addr,
+		// TODO: what is default? namespace? partition?
+		fmt.Sprintf("%s.default.%s.external", upstream.ID.Name, upstream.Peer),
+		"HEALTHY",
+		1,
+	)
+}
+
+type testingT interface {
+	require.TestingT
+	Helper()
+}
+
+// does a fortio /fetch2 to the given fortio service, targetting the given upstream. Returns
+// the body, and response with response.Body already Closed.
+//
+// We treat 400, 503, and 504s as retryable errors
+func (a *Asserter) fortioFetch2Upstream(
+	t testingT,
+	client *http.Client,
+	addr string,
+	upstream *topology.Upstream,
+	path string,
+) (body []byte, res *http.Response) {
+	t.Helper()
+
+	url := fmt.Sprintf("http://%s/fortio/fetch2?url=%s", addr,
+		url.QueryEscape(fmt.Sprintf("http://localhost:%d/%s", upstream.LocalPort, path)),
+	)
+
+	req, err := http.NewRequest(http.MethodPost, url, nil)
+	require.NoError(t, err)
+
+	res, err = client.Do(req)
```

Comment:
The latter half of this is not in a `retry.Run`. This is because it was already  being called from a `retry.Run` and the `*testing.T` and `*retry.R` values were getting mixed up leading to retry-suppression.

A few of the `t` vars below have been properly replaced with `r`.